Mount snapshots on Windows

gabriel-samfira opened the pull request:
This PR is just a rebase of #4419.

@TBBle is currently on sabbatical so I will try to move the two PRs (this one and the one in https://github.com/microsoft/hcsshim/pull/1637), forward.

All needed context for this PR is available in #4419 .

In addition to the great work done in #4419, this PR makes use of the Bind Filter in Windows, which gives us functionality similar to what we have on Linux via bind mounts. It allows us to avoid some Windows special cases.

  * Depends on: https://github.com/microsoft/hcsshim/pull/1637
  * ~~Depends on: https://github.com/containerd/continuity/pull/212~~
  * Depends on: https://github.com/containerd/continuity/pull/219
  * Depends on: https://github.com/microsoft/go-winio/pull/274

Note: This PR will fail to build without the above mentioned dependency. Marking it as draft for now. Also, there is still some cleanup I need to do (ex: remove unused functions), which will be addressed shortly.

gabriel-samfira commented:
Dropped commit https://github.com/gabriel-samfira/containerd/commit/6dae7f9bd046fff48780e9c25e16486f84b3a569 as we're using the bind filter, which doesn't use junction points. Thus, this issue should not manifest.

Also, with the change to continuity (https://github.com/containerd/continuity/pull/212), we'll also get a proper diff between layers.

gabriel-samfira commented:
The snapshot tests require the ```SeBackupPrivilege``` and ```SeRestorePrivilege``` to have access to system files inside the container layer mount points.

gabriel-samfira commented:
This commit: https://github.com/containerd/containerd/pull/8043/commits/03ffa12b09807cd9321d2635089ebcbb3d6b5bd4

Adds two new files, one of which is generated code via ```mksyscall```. I can add the headers to one of them, but the other one is generated. Any tips on how to handle this situation? Should this be moved in some other package?

@helsaawy Do you think something like this should be added to ```go-winio```? Any other suggestions are welcome.

dcantah commented on mount/mount_windows.go:
```diff
@@ -26,8 +27,30 @@ import (
+
 // Mount to the provided target.
 func (m *Mount) mount(target string) error {
+	readOnly := false
+	for _, option := range m.Options {
+		if option == "ro" {
+			readOnly = true
+			break
+		}
+	}
+
+	if m.Type == "bind" {
+		if err := ApplyFileBinding(target, m.Source, readOnly); err != nil {
+			return fmt.Errorf("failed to bind-mount to %s: %w", target, err)
+		}
+		return nil
+	}
+
```
This is awesome!!!

dcantah commented:
@gabriel-samfira Ping me when this is ready!

dcantah commented on mount/bind_filter_windows.go:
```diff
@@ -0,0 +1,327 @@
+package mount
```
This makes me wonder if it'd be better suited to be in go-winio. We can lock down the API there, and then hcsshim and here could all consume from it.

gabriel-samfira replied:
Yup. Already proposed a ```go-winio``` PR here: https://github.com/microsoft/go-winio/pull/274. Need to write some tests today. Once that is in, I'll update the module and remove these 2 files.

dcantah replied:
Oh sweet! Should've checked 😄

gabriel-samfira commented:
> @gabriel-samfira Ping me when this is ready!

You bet! The more reviews and scrutiny this gets, the less stressed I'll be that I made a silly mistake somewhere which will blow up later :smile: .

gabriel-samfira commented:
Hi @dcantah !

If you have the time, feel free to review this branch. Sadly, I didn't get all the "metadata" files in the previous PR to continuity. I opened a new one here: https://github.com/containerd/continuity/pull/219.

In this branch, I have a temporary hack that points to my continuity branch, but other than that, this should be ready to review. Have a look when you get a chance!

Thanks!

dcantah commented:
Woo! I'll definitely carve out some time this week

gabriel-samfira commented:
Integration tests seem to have passed! :smile: . Well...everything aside from a couple of ```Vagrant BOX``` tests.

https://github.com/containerd/containerd/actions/runs/4297996002

gabriel-samfira commented:
> Oh, we aren't seeing the test failures that lead to [bd51841](https://github.com/containerd/containerd/commit/bd51841caa9180fb3f17d1575f96b5e3b3b3995c)? Awesome!

If it was "File in use" it may have been an older bug. I remember fixing one like that here:

https://github.com/microsoft/hcsshim/pull/1249

Not sure if it is the same one you were seeing back in 2021 😄 .

Increasing the timeout sounds prudent (we had to do it before on Windows). Will probably do that in a separate PR.

gabriel-samfira commented:
> Oh. A quick look at the test run suggests we might need to bump the Windows Integration test-run time-limit, it's currently 50m, and one of the four runs hit this in a very late stage (critest execution). The other three are also really close, so I don't think this was a failure or problem, it's just that with the new tests enabled from the snapshot suite, the expected runtime is longer. I'm guessing 60 minutes would be enough, but to avoid false-negatives, perhaps 90m?

@aznashwan was kind enough to open a PR for this here: https://github.com/containerd/containerd/pull/8183

Thanks @aznashwan !

gabriel-samfira commented:
Marking this as ready for review, as there is just one simple PR it depends on, and it would be useful to get more eyes on this sooner rather than later.

TBBle commented:
> > Oh, we aren't seeing the test failures that lead to [bd51841](https://github.com/containerd/containerd/commit/bd51841caa9180fb3f17d1575f96b5e3b3b3995c)? Awesome!
> 
> If it was "File in use" it may have been an older bug. I remember fixing one like that here:
> 
> [microsoft/hcsshim#1249](https://github.com/microsoft/hcsshim/pull/1249)
> 
> Not sure if it is the same one you were seeing back in 2021 😄 .

That won't be it, as the one I was hitting was triggered by running the snapshotter test suite in parallel mode (basically the 128 layers test alongside some other test would trigger it about 60% of the time on Server 2019, passing just often enough to give me false hope), and so there were no containers/shims involved. I thought I'd fixed it when I found an unclosed handle in an old version of the "base layer manipulation" PR, and when I pulled that in, it reproduced on Windows 2022 as well. >_<

We also occasionally saw the same error happen during other test runs in containerd main, although the bug you fixed above _might_ have been that as those tests would probably have involved containers, as outside this PR, the Windows snapshotter is not being run through the snapshotter testsuite, only in trivial "Import/Export" tests and the container-running test suite.

gabriel-samfira commented:
> We also occasionally saw the same error happen during other test runs in containerd main, although the bug you fixed above _might_ have been that as those tests would probably have involved containers, as outside this PR, the Windows snapshotter is not being run through the snapshotter testsuite, only in trivial "Import/Export" tests and the container-running test suite.

I've been running ```sh.exe -c "EXTRA_TESTFLAGS=-test.run=TestSnapshotterClient make integration"``` locally for a while now without issues. Most errors have been in regards to differences due to "metadata" files, and a few that were due to not having access to read system files. Both of which have been addressed. Whatever bug was triggering that was probably fixed :crossed_fingers:

gabriel-samfira commented:
> Edit: Thinking about it, since the `check128LayersLockstep` test in that reliable repro link is self-contained, it might be worth adding it into this PR and making sure it passes on CI. Maybe not worth adding to containerd main though, since it's a trimmed version of a test that already exists.

The check128LayersLockstep should run as part of this PR. Test output is suppressed, but it should run, both as part of ```make integration``` and as part of ```make test```.

TBBle commented:
Sorry, `check128LayersLockstep` wasn't ever in this PR series, it was on my "Hack until I can nail down this bug" branch derived from it, you can see it in all its glory in https://github.com/TBBle/containerd/commit/50052cbca4ff98ddbba5036cefcddd4a64080ea8#diff-0c729c2c9e5ed5e5402da537dceb59655a78b9f6ac17cb7a9a17a94232403b15R1146-R1191.

It's basically the minimised version of `128LayersMount` that still reliably triggered the bug, and then rigged up to be run in 10 parallel instances with itself, which gave me basically a 100% repro rate if I ran just that suite. (You can see in that commit that I had disabled the rest of the snapshotter suite so I'm guessing I was using the same `-test.run=TestSnapshotterClient` parameter you mentioned above to trigger it.

Per https://github.com/TBBle/containerd/commit/46efff17508c81b49d931b960fc6ff7972d43287, commenting out the `mount.All` would stop reproduction, so whatever was going on was somewhere inside the `mount.All` (or paired unmount) call. Looking at the history of that branch, I don't appear to have ever nailed down what specifically inside that was actually reproducing the problem.

Thinking back, I don't think it occurred to me to try removing just the volume mount point manipulation, so this being fixed in-passing by switching to bind-mount is definitely feasible, but doesn't given any insight into why I couldn't reproduce this with hcsshim directly, unless I wasn't doing the mounting in my attempt to reproduce it there.

dcantah commented on diff/windows/windows.go:
```diff
@@ -320,10 +320,13 @@ func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
 		// This is a special case error. When this is received the diff service
 		// will attempt the next differ in the chain which for Windows is the
 		// lcow differ that we want.
+		// TODO: Is there any situation where we actually wanted a "bind" mount to
+		// fall through to the lcow differ?
```
afair, this was just to handle unpacking linux images from the `Apply` path. lcow has a type of lcow-layer so we should be fine

TBBle replied:
Yeah, if it's a generic feature, then I think the current approach of "Explicitly allow-list mount types we understand, and let others fall through to whatever can handle it" is better. Technically, we probably should reject an upper layer `bind` mount the same way, right now it'll come though as 'Invalid Argument' (no parents) and fail rather than try the next differ. (I _think_ that logically a parentless layer will always be a bind mount and vice versa here by dint of the snapshotter implementation, but am not immediately certain that outside forces won't break that assumption.)

dcantah commented on integration/client/go.mod:
```diff
@@ -21,13 +21,14 @@ require (
 	golang.org/x/sys v0.5.0
 )
 
+require github.com/Microsoft/go-winio v0.6.1-0.20230228163719-dd5de6900b62
```
Somehow this ended up all by its lonesome and not with the other required deps

dcantah commented on mount/mount_windows.go:
```diff
@@ -46,22 +72,42 @@ func (m *Mount) mount(target string) error {
 	}
+	defer func() {
+		if err != nil {
+			hcsshim.DeactivateLayer(di, layerID)
```
This and Unprepare always scare me in terms of what they could fail with, so wanna grab the error and log here and same for on line 86? Up to you, it would bring me peace of mind 😭

dcantah commented on mount/mount_windows.go:
```diff
@@ -46,22 +72,42 @@ func (m *Mount) mount(target string) error {
+		}
+	}()
+
+	// Add an Alternate Data Stream to record the layer source.
+	// See https://docs.microsoft.com/en-au/archive/blogs/askcore/alternate-data-streams-in-ntfs
+	// for details on Alternate Data Streams.
+	if err = os.WriteFile(filepath.Clean(target)+":"+sourceStreamName, []byte(m.Source), 0666); err != nil {
+		return fmt.Errorf("failed to record source for layer %s: %w", m.Source, err)
+	}
```
Clever! @TBBle

dcantah replied:
Although, kind of annoying we even need to do this

TBBle replied:
This wasn't my idea, I cadged it from the even older "Mount Windows" PR #2287.

It is a shame that there's no way to get the _layer_ being unmounted from either containerd (Unmount only takes the mount point name, because that's how the umount syscall works on Linux...) or from the OS (because HCS's XXXLayer API lacks a way to lookup layer details from the WCIFS virtual volume; I don't recall if the computestorage API changed this).

Not to say this isn't a reasonable solution. It even has the nice effect that deleting the mount point folder automatically removes the ADS data too so there's no out-of-sync data feasible, and makes the relevant data available to some kind of inspection/recovery tool if needed, e.g. to unblock layer deactivation in the case of containerd death.

Hopefully we never hit a problem with this from some tool that goes weird in the face of ADS data, like Go's filepath.walk does in the face of junction points.

dcantah commented on mount/mount_windows.go:
```diff
@@ -85,25 +131,52 @@ func (m *Mount) GetParentPaths() ([]string, error) {
-		return fmt.Errorf("failed to deactivate layer %s: %w", mount, err)
+
+	if err := bindfilter.RemoveFileBinding(mount); err != nil {
+		if errno, ok := errors.Unwrap(err).(syscall.Errno); ok && errno == windows.ERROR_INVALID_PARAMETER || errno == windows.ERROR_NOT_FOUND {
```
Wait, can we do `errors.Is(err, windows.ERROR_INVALID_PARAMETER) || errors.Is(err, windows.ERROR_NOT_FOUND)`?

TBBle replied:
I think so, yeah. We couldn't for the hcsshim-supplied errors (see https://github.com/microsoft/hcsshim/issues/960) but go-winio's bindfilter functions are just using `fmt.Errorf` with `%w` for the errors coming from mkwinsyscall, so `errors.Is` should work naturally. AFAIR, the `windows.ERROR_*` constants do work correctly with `errors.Is`.

dcantah commented on snapshots/windows/windows.go:
The first case is kinda wonky now (to me at least), could we re-order this so one of the cases isn't just housing a comment

TBBle replied:
The boolean logic got ugly when I tried that before. We could invert the first test and push everything a level deeper, or copy the inverted test into each following if/elseif.

Probably the least-ugly option would be to push all the logic in the second and third branches out into a new "ensureDataStorage" function (or two functions...). Although that loses a nicety of this PR that the `else` branch is unmodified, which helps keep the diff readable.

As a quick fix, looking at the current logic, that first branch could probably `return nil`, since we're already inside a lambda which is in-effect `storage.CreateSnapshot` followed by `ensureDataStorage`, and this if-else block is followed by `return nil` already.

In fact, if the `if` and `else if` ended with `return nil`, then the `else` block can be _unindented_ a level in-place. That's actually pretty nice, but again costs the PR a modicum of "did this code change?" readability. (Probably a worthwhile sacrifice, and turning on ignore-whitespace would resolve that anyway). As long as we won't ever need to do anything common to these directories after this step, I guess.

dcantah commented:
Should have some spare hours tomorrow, but there's a couple bits of feedback I left here to get started!

gabriel-samfira commented:
> Should have some spare hours tomorrow, but there's a couple bits of feedback I left here to get started!

Thank you, @dcantah ! I think I addressed the current comments. Can't wait for the next round :smile:

TBBle commented on mount/mount_windows.go:
```diff
@@ -74,16 +75,21 @@ func (m *Mount) mount(target string) error {
 		if err != nil {
-			hcsshim.UnprepareLayer(di, layerID)
+			if layerErr := hcsshim.UnprepareLayer(di, layerID); layerErr != nil {
+				log.G(context.TODO()).WithError(layerErr).Error("failed to deactivate layer")
```
Nit: "failed to unprepare layer". I'd also suggest adding "during mount failure cleanup" or something to these, so they're easily distinguishable from the same function calls failing in Unmount.

gabriel-samfira commented:
CC: @helsaawy

If you could have a quick look over this PR, it would help a lot. Feel free to tag anyone else that may have input.

gabriel-samfira commented:
Updated continuity to latest commit and removed the hack that redirects to my branch. Tests seem to pass, except the Vagrant box ones that do not seem to be related.

dcantah commented:
Trying to find time for a second look, haven't forgotten about this! Also, seems there's a merge conflict

gabriel-samfira commented:
> Also, seems there's a merge conflict

Ohh! That's new. Wasn't there this morning. Will rebase asap

gabriel-samfira commented:
Well, that didn't last long 😄 . I'll have to rebase again.

gabriel-samfira commented:
CC: @kevpar

If you could have a quick look over this PR, it would help a lot.

dcantah commented on integration/client/snapshot_windows_test.go:
```diff
@@ -0,0 +1,40 @@
+	// The SeBackupPrivilege and SeRestorePrivilege gives us access to system files inside the container mount points
+	// (and everywhere on the system), without having to explicitly set DACLs on each location inside the mount point.
+	if err := winio.EnableProcessPrivileges([]string{winio.SeBackupPrivilege, winio.SeRestorePrivilege}); err != nil {
+		t.Error(err)
```
Lets t.Fatal here if this is necessary for the test suite to work

dcantah commented on snapshots/windows/windows.go:
Honestly this is how it should've been probably, I like early returns

dcantah reviewed (approved):
This seems like one of those types of changes where you put some extreme faith in the tests 🤓 LGTM seeming as how the windows CI is smiling

gabriel-samfira commented:
> This seems like one of those types of changes where you put some extreme faith in the tests

That kind of happens with all code tbh :smile:. Luckily, the tests abuse this functionality quite a bit (bind filter included). Seems to not blow up, and appears to work as intended.

gabriel-samfira commented:
GitHub/Azure seems to be having an issue today. Failures due to timeouts.

kzys commented on integration/client/snapshot_unix_test.go:
```diff
@@ -0,0 +1,35 @@
+import (
+	"testing"
+
+    . "github.com/containerd/containerd"
```
Nit: you don't have to `.` import containerd. Do you?

TBBle replied:
integration/client has its own go.mod, so I suspect you do. Certainly the existing files I checked do this too.

gabriel-samfira replied:
> integration/client has its own go.mod, so I suspect you do. Certainly the existing files I checked do this too.

Quite right. Completely forgot about the ```go.mod``` in integration/client.

kzys commented on snapshots/testsuite/testsuite.go:
```diff
@@ -891,19 +902,19 @@ func check128LayersMount(name string) func(ctx context.Context, t *testing.T, sn
-			fstest.CreateDir("/ADDHERE", 0755),
-			fstest.CreateDir("/ONLYME", 0755),
-			fstest.CreateFile("/ONLYME/bottom", []byte("bye!\n"), 0777),
+			fstest.CreateDir("/addhere", 0755),
```
Do you need to change them?

TBBle replied:
Yes. Windows has the amusing behaviour that an all-caps file is assumed to be a DOS legacy, and recased before writing. So it's either change the case, or make the comparison code case-insensitive. See https://github.com/containerd/containerd/pull/8043/commits/a9bf2f30dcf4eec768864179b82785cbad216355

gabriel-samfira replied:
> Yes. Windows has the amusing behaviour that an all-caps file is assumed to be a DOS legacy, and recased before writing. So it's either change the case, or make the comparison code case-insensitive. See [a9bf2f3](https://github.com/containerd/containerd/commit/a9bf2f30dcf4eec768864179b82785cbad216355).

The mine fields we tend to create when we can't stop ourselves from adding code that tries to be smarter than it needs to be. Thanks for the context! TIL :smile:

kzys commented on snapshots/testsuite/testsuite.go:
```diff
@@ -944,7 +957,7 @@ func check128LayersMount(name string) func(ctx context.Context, t *testing.T, sn
 			}
 
-			if err := fstest.CheckDirectoryEqual(preparing, flat); err != nil {
+			if err := fstest.CheckDirectoryEqual(flat, preparing); err != nil {
```
I'm inclined to keep them as is, assuming they are semantically same.

fuweid replied:
@gabriel-samfira This patch is not small. So I think we should not touch un-related code. It can be handled in the followup.

gabriel-samfira replied:
It is somewhat related. Without this, the test suite fails. The ```CheckDirectoryEqual()`` function needs the first folder to be the "parent" folder and the second one to be the "snapshot" of the parent. Reversing them breaks the comparison when ```metadataFiles``` are involved.

I can undo this, but then I'd also have to disable the tests, otherwise they will never pass.

TBBle replied:
This change only touched three lines, according to https://github.com/containerd/containerd/pull/8043/commits/98236a851c001d00218121234c1106f213b80711.

samuelkarp commented:
/ok-to-test

fuweid commented on snapshots/testsuite/testsuite.go:
```diff
@@ -817,6 +823,7 @@ func checkSnapshotterViewReadonly(ctx context.Context, t *testing.T, snapshotter
 	if err := os.WriteFile(testfile, []byte("testcontent"), 0777); err != nil {
 		t.Logf("write to %q failed with %v (EROFS is expected but can be other error code)", testfile, err)
 	} else {
+		testutil.Unmount(t, viewMountPoint)
```
I think we can use the following pattern. WDYT?

```go
err := os.WriteFile(testfile, []byte("testcontent"), 0777)
testutil.Umount(t, viewMountPoint)
if err != nil {
     t.Logf("write to %q failed with %v (EROFS is expected but can be other error code)", testfile, err)
} else {
     t.Fatalf("write to %q should fail (EROFS) but did not fail", testfile)
}

fuweid reviewed (commented):
Since there are some changes in snapshot testsuite, I would like to ask @dmcgowan to take a look on this.

TBBle commented on integration/client/snapshot_windows_test.go:
```diff
@@ -0,0 +1,42 @@
+	if err := winio.EnableProcessPrivileges([]string{winio.SeBackupPrivilege, winio.SeRestorePrivilege}); err != nil {
+		t.Fatal(err)
+	}
+	defer winio.DisableProcessPrivileges([]string{winio.SeBackupPrivilege, winio.SeRestorePrivilege})
```
The tests inside `testsuite.SnapshotterSuite` run in parallel mode and I assume they will all complete before this defer executes. However I guess other tests could overlap and have an unexpected privilege in some runs but not others.

It's possible we can rejig this to take a per-thread privilege inside the relevant test routines, if we can rely on them staying on the same thread.

Otherwise, we could split the test suite by tag, I guess, so that tests that require backup/restore privileges are separated.

This is made slightly not-so-bad by the existing second integration test run with parallel disabled, so unless extra privileges cases a _failure_, a false-success should be caught on the second run.

gabriel-samfira replied:
Let me try thread local privilege escalation.

gabriel-samfira commented on mount/mount_windows.go:
```diff
@@ -46,22 +73,49 @@ func (m *Mount) mount(target string) error {
 		return fmt.Errorf("failed to activate layer %s: %w", m.Source, err)
 	}
+	defer func() {
+		if err != nil {
```
It would clearly highlight which err we're handling here if we explicitly define it as a named return. Changing.

gabriel-samfira reviewed (commented):
@fuweid Thank you for the review! Testing out the thread-local privilege escalation and replacing the process privilege escalation if it works.

helsaawy commented on mount/mount_windows.go:
```diff
@@ -17,17 +17,44 @@
 
 // Mount to the provided target.
 func (m *Mount) mount(target string) error {
+	readOnly := false
+	for _, option := range m.Options {
+		if option == "ro" {
+			readOnly = true
+			break
+		}
+	}
```
We use this here and in the snapshotter, can we add a `(m *Mount) ReadOnly() bool` function?

TBBle replied:
Frankly, "search the `mount(8)` command-line options array for 'ro'" approach has always bugged me (but not enough to propose doing things differently, given that this is an API type). _Hiding_ it would be a nice first step.

gabriel-samfira replied:
@helsaawy Yes. Definitely. Changing now.

helsaawy reviewed (approved):
Windows specific bits LGTM

TBBle commented on mount/mount_windows.go:
```diff
@@ -17,17 +17,44 @@
+		}
+	}
+
+	if m.Type == "bind" {
+		if err := bindfilter.ApplyFileBinding(target, m.Source, readOnly); err != nil {
+			return fmt.Errorf("failed to bind-mount to %s: %w", target, err)
+		}
+		return nil
+	}
+
```
I'm not sure what you mean. We support two mount types, `bind` (for a base-layer's Files/ directory, trivially handled here and early-out'd) and `windows-layer`, the fall-through, and everything else is rejected by the `if m.Type != "windows-layer"` check immediately after.

helsaawy replied:
I think I confused this with the snapshotting code below, my bad

TBBle commented on snapshots/windows/windows.go:
```diff
@@ -312,12 +314,19 @@ func (s *snapshotter) mounts(sn storage.Snapshot) []mount.Mount {
+	if len(sn.ParentIDs) == 0 {
+		// A mount of a parentless snapshot is a bind-mount.
+		mountType = "bind"
+		// If not being extracted into, then the bind-target is the
+		// "Files" subdirectory.
+		if !strings.Contains(key, snapshots.UnpackKeyPrefix) {
+			source = filepath.Join(source, "Files")
+		}
```
Maybe this logic (appending `Files`) should actually live in the mount code?

That would make this key-check unnecessary, as the assumption underlying the current code's correctness is that imports never mount the directory, and the only users of the Source key are importing and mounting. It'd be nice to remove that assumption, since for example I've just been reminded we pass the `Mount` object out into the runtime shim, so avoiding surprises there seems valuable.

That'd also mean one fewer places that rely on the snapshot _key name_, which I'm not super-fond of anyway.

gabriel-samfira replied:
Moved this append in the ```mount()``` function. Running tests now. If all is well, I'll update the PR. Good catch!

gabriel-samfira replied:
Seems ok.

fuweid commented on snapshots/windows/windows_test.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+func TestWindows(t *testing.T) {
+	testutil.RequiresRoot(t)
```
It is no-op function. Does it work?

TBBle replied:
Huh, interesting that this is a no-op. Maybe we should implement that some day (to check for elevation though, not Administrator), and see if the test-suite can pass as non-root (and work out if anything's needlessly skipped when that's done).

That said, the Windows snapshotter absolutely relies on being run elevated (since the differ grants itself privileges, and the tests need to mount the results for validation), so this seems like the right thing to call here. The Native Snapshotter test suite makes the same check for (approximately) the same reasons.

fuweid replied:
I think the CI is running with root role. So it can pass. Perhaps we can add check and abort if it is not root. It can be handled in the follow-up.

TBBle replied:
Oh yes, containerd CI runs as an elevated account on Windows, that's [the default for GHA-owned VMs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#administrative-privileges), and for the Azure-hosted VMs see, e.g., [the VM creation command](https://github.com/containerd/containerd/blob/988ee8ffef1e756039fe05539be3398e70ed3e0c/.github/workflows/windows-periodic.yml#L98).

I don't think it's worth trying to change this, either. Windows CI runs both `test` and `root-test` together right now, and I don't see specific value in splitting that.

So making `RequiresRoot` skip tests when non-elevated (as the Unix version does) would simply give a better UX to developers who run the test suite on Windows from a non-elevated account, either by accident or because the individual test the user is interested in does not require 'root'. I agree _this_ is probably worth doing independently of this PR.

TBBle commented on mount/mount_windows.go:
```diff
@@ -38,18 +38,22 @@ var (
+func (m *Mount) mount(target string) (retErr error) {
 	if m.Type == "bind" {
-		if err := bindfilter.ApplyFileBinding(target, m.Source, readOnly); err != nil {
+		if !strings.EqualFold(filepath.Base(m.Source), "Files") {
+			m.Source = filepath.Join(m.Source, "Files")
+		}
+		if err := bindfilter.ApplyFileBinding(target, m.Source, m.ReadOnly()); err != nil {
```
I'd prefer to not modify `m` in-place (which I assume is why we check for a second `Files`), partly because this function really has no place modifying `m`, and partly because _this package_ should not be assuming that some person hasn't unexpectedly created a base layer in a directory named `Files`.

That would do the wrong thing with this check; the built-in Windows snapshotter will never do that, but an alternative implementation in an external snapshotter plugin might?

So I expect that
```
if err := bindfilter.ApplyFileBinding(target, filepath.Join(m.Source, "Files"), m.ReadOnly()); err != nil {
```
would work fine, unless I'm overlooking something.

gabriel-samfira replied:
Okay. If I understand this correctly, in the case of bind mounts we need to:

  * Mount the layer's ```Files``` folder in the target
  * Make sure that ```Files``` folder actually exists. This should happen in ```hcsshim``` with the new base layer code. It should be fine not to check for the existence of that folder.

gabriel-samfira replied:
made the change

gabriel-samfira commented on mount/mount_windows.go:
```diff
@@ -17,17 +17,48 @@
-func (m *Mount) mount(target string) error {
+func (m *Mount) mount(target string) (retErr error) {
+	if m.Type == "bind" {
+		if !strings.EqualFold(filepath.Base(m.Source), "Files") {
```
@TBBle I believe this is what you meant. Lemme know if I misread :smile:

TBBle replied:
Heh, we crossed comments: https://github.com/containerd/containerd/pull/8043/files/4012c1b853adf637667ad62c74c3fa61b3f5e22e..b837c141def583d3e7add3015bd16623ed28e221#r1154603885

TBBle commented on diff/windows/windows.go:
```diff
@@ -332,6 +335,22 @@ func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
+		} else if len(parentLayerPaths) == 0 {
+			return "", nil, fmt.Errorf("unexpected windows-layer View with no parent: %w", errdefs.ErrInvalidArgument)
+		}
+		// Ignore the dummy sandbox.
```
I just noticed that this comment's [partner in `createSnapshot`](https://github.com/containerd/containerd/pull/4419/files#diff-b6ae335f862c747466f1bc00e89eb1ef7cd9589bcc89ec57953c9c20125b3752R425-R427) has been lost as part of the refactoring to use BindFilter.

We also no-longer ignore the rootfs-size labels in the read-only case, although I hope no one passes such labels into View creation...

It might be worth adding a comment [at that point in the snapshotter](https://github.com/containerd/containerd/pull/8043/files#diff-b6ae335f862c747466f1bc00e89eb1ef7cd9589bcc89ec57953c9c20125b3752R420) that when generating mounts for Views, we still create a scratch layer, because we can't actually mount a stack of layers on Windows unless the top one is a scratch layer, and noting that the differ knows to ignore the scratch layer when comparing layers in such mounts. This comment could then directly refer to that comment, since if you're not reading the PR, it's not clear what "Ignore the dummy sandbox" is referring to.

To be clear, I am _assuming_ BindFilter still requires a top-most scratch layer to mount. In theory a locally-created layer is both scratch and read-only (we don't delete sandbox.vhdx as the hcs layer-differ consumes that from the top layer, not the exported format) so such a layer could perhaps be mounted directly as a view using BindFilter's read-only flag, and no dummy sandbox. We'd need to communicate that through to the differ somehow (or make the mounter more complex and handle it there), so consider such an exploration as a possible future optimisation that's not even strongly justified. (And merely mounting the sandbox may mutate it, which is pretty non-ideal.)

TBBle replied:
Huh, for some reason I had the idea it could actually be given the list of layers to mount, and take care of WCIFS activation internally. Sorry, ignore that last paragraph then. ^_^

TBBle commented on snapshots/windows/windows.go:
```diff
@@ -312,12 +314,14 @@ func (s *snapshotter) mounts(sn storage.Snapshot) []mount.Mount {
+	source := s.getSnapshotDir(sn.ID)
+	parentLayerPaths := s.parentIDsToParentPaths(sn.ParentIDs)
+
+	mountType := "windows-layer"
+
+	if len(sn.ParentIDs) == 0 {
+		// A mount of a parentless snapshot is a bind-mount.
+		mountType = "bind"
```
Because the mount code does the `Files/` sub-directory thing, is there any chance that calling this a `bind` mount (which seemed cute when I first faked it with symlinks) is going to confuse anything/anyone? Something like `windows-base-layer` would be nice and distinct.

I wonder if we could just use `windows-layer` for this too and count the number of parent layers instead? The only time we distinguish the two mount-types, we also have a call to m.GetParentPaths or access to its count, and one of those times, we're erroring for the "bind with parents" and "non-bind without parents" cases anyway. So it _seems_ like it would work, on inspection.

gabriel-samfira replied:
I am inclined to leave this as a ```bind``` mount. In effect, it walks and it talks like a bind mount, and it is far more intuitive to understand to anyone coming from the *nix world than a special case of ```windows-layer```. Coincidentally, the filter we use to make this happen is also called ```BindFilter``` :smile:.

I don't feel particularly strongly about this, so if you feel like this is something that needs to be changed, I will make the changes.

TBBle replied:
That was actually why I was wondering if it should be changed, because it's _close_ to a Linux bind mount, but by handling `Files/` the way we do now, I was wondering if that was a false equivalence, since we currently assume/expect that the mount source being passed to it is a WCOW base layer, not an arbitrary directory.

That would allow us to reserve `bind` for an _actual_ arbitrary directory bind-mount that a caller might request with a hand-constructed `Mount` object, which _is_ what we had implemented before ebf9906ed6336c95ba9cc6be74d83ce858e30958, on reflection.

So maybe ebf9906ed6336c95ba9cc6be74d83ce858e30958 was wrong-headed, and a base-layer-only Active mount _should_ be a trivial bind-mount to `Files/`, with a special case for unpacking which returns a mount which will do the wrong thing if mounted? The special case for unpacking in `createSnapshot` already "returns a mount which will do the wrong thing if mounted" for non-base layers, so this isn't as bad as my phrasing makes it sound; it'd be consistent, at least. I had a quick poke around with GitHub, and the only places in containerd that use the unpacking key are local operations that immediately `archive.Apply` and `snapshot.Commit`, so the only place this has risk is users directly calling the snapshotter API.

Anyway, you're closer to working with this code than I am now, so I'm fine with whatever you feel minimises surprise/risk etc. My feeling is that the current state is caught between two worlds, and we should either move the append of `Files/` back into the snapshotter (restoring "Bind mount does what it says on the box") or we push through with "`windows-filter` is the only mount-type we know, and it knows how to handle both base and non-base layers".

I also feel that if we're going to go with "bind", such mounts should not have `parentLayersOption` included in the `Mount` object, since by snapshotter definition that's empty for a bind mount, and if it's not empty, we'd reject it at mount-time anyway. (The presence of `parentLayersOption` in bind mounts is what opened this question up for me in the first place.)

gabriel-samfira replied:
I'll have to think about this for a bit.

The ability to "bind" mount is already (sort of) used in ```buildkit``` here: https://github.com/moby/buildkit/blob/master/snapshot/localmounter_windows.go#L30-L52

Some clients already expect to be able to "bind" mount and have that work in a certain way. If we decide to do away with "bind" mounts now, until we can decide on a more accurate implementation, the ```buildkit``` code can be adapted to Windows, but would introduce additional code paths to account for the differences.

The way Windows layers work make it a bit tricky. From a client's perspective, the ```Files``` folder is the actual "```rootfs```". If I understand things correctly, even for a ```Hyper-V``` isolated container (Xenon), the same rule applies. The difference is, we also have the ```UtilityVM``` with it's own files needed to boot up a VM, in which the actual container is started, and for which the ```Files``` folder becomes the ```rootfs``` of the container. Sort of like Kata containers if we think about it, but in this case, the plumbing that Kata adds is included in the actual layer on Windows.

So the code as it is now, enables clients to mutate or view the ```rootfs``` of a container, completely ignoring any support files the layer contains. The same thing happens on *nix systems, as those don't really have all the supporting files for ```UtilityVMs```, overlay hives or the needed BCD hives to actually "boot" a ```UtilityVM``` or a container.

As far as the clients are concerned, the API is similar to what they get on Linux. If I understand your concerns correctly, there remains the issue where we want to mount a layer for the purpose of unpacking an actual image into it. I will have to follow that code on Monday and see how that happens. I haven't looked at that much.

As for clients, I am trying to think of a reason a client might want to bind mount the actual layer, as opposed to the "```rootfs```" of the containers that resides in that layer. Mounting a ```windows-layer``` on a host will also expose the ```Files``` folder. So it doesn't seem like clients should be concerned with the internal plumbing of how a Windows container layer is handled. Further more, I would be inclined to prevent direct access to the base layer for clients, as there is no real reason to mess with internals as a client and that might actually break things in unexpected ways, if the "support" files are tampered with.

For now, I will revert this PR pre https://github.com/containerd/containerd/commit/ebf9906ed6336c95ba9cc6be74d83ce858e30958. If you're willing, we can have a broader discussion on slack, at your convenience, sometime next week. But only if you can spare ~1 hour or so. No worries if not.

gabriel-samfira replied:
> I also feel that if we're going to go with "bind", such mounts should not have `parentLayersOption` included in the `Mount` object, since by snapshotter definition that's empty for a bind mount, and if it's not empty, we'd reject it at mount-time anyway.

I've added: https://github.com/containerd/containerd/pull/8043/commits/ca5605b4a208d9d5c809813adbec6b2227ab6c23

TBBle replied:
I haven't gone through the rest (on my way to bed) but...

> The ability to "bind" mount is already (sort of) used in `buildkit` here: https://github.com/moby/buildkit/blob/master/snapshot/localmounter_windows.go#L30-L52

I'll note that I'm partly responsible for that code, and it contains at least one complete misunderstanding by me of how WCOW layer mounts work, right at the start of this adventure. (But not the bit you're talking about here, that's common with the Linux localmounter)

It does emphasise my point somewhat that if we say "bind" as the type, then `Mount.source` _better_ be what will appear at the target location, since BuildKit does this "Bypass rw bind mounts" trick. What we have now will break that, and if we revert to the state we were in before (which on skimming, I see you did), then we'll be in a better place.

I _think_ that once we have landed this (whether we use "bind" and BuildKit's bypass works, or "windows-filter" and BuildKit trustingly calls `mount.All` to create a sparkling symlink), the BuildKit localmounter code for non-Windows should also be usable on Windows allowing us to unify that split. It's possible `MNT_DETACH` will bite us here though, I don't recall if `mount.Unmount` ignores options on Windows or enforces 0.

----

Okay, I did read the rest of the comment, and generally agree with you. I will note that for non-base layers (`windows-layer`) mounts, the `Mount.Source` also has all the metadata etc. visible. So if we call the base layers `windows-layer`, then `Mount.Source` exposing the metadata is consistent. I agree that if we call it a `bind` mount, we should not expose the metadata, just point at the contained files, like you'd see if you actually mounted it.

The other user of the Mount structure outside of containerd is the runtime shims; hcsshim of course needs to see the metadata, to launch a container, but we'll never give it a base-layer mount as it cannot execute that (it needs a sandbox on top, AFAIK, although Host Process Containers work under very different limitations).

The most-conservative route would be to only return `windows-layer` mounts (i.e. the state of the world before this PR), and hence even though a base-layer mount would happen to be a trivial bind-mount, inspecting the Mount object wouldn't tell you that. This would let us reduce the "unpack" behaviour differences, and prevents the BuildKit "rw-bind-mount" bypass.

However, if you're confident that for a base-layer snapshot, we can produce a Mount object that looks just like a bind-mount on Linux, then that's not an issue. That's the state I think we were in before https://github.com/containerd/containerd/pull/8043#discussion_r1153930966, and at this point, I don't have any cases where it'll go wronger than the existing code that skips sandbox.vhdx creation, and your testing hadn't shown up any issues with BuildKit here.

gabriel-samfira replied:
Apologies for the wall of text :smile:

I've been thinking about the issue of mounts and how we could clarify what we want to achieve here. A bit of context first, so we don't get lost in details. You are aware of most of this context, as you've basically written this :smile:, but for the sake of keeping the conversation on track, I think it's worth adding here.

## Arbitrary base layers

This feature came (mostly) out of the need for buildkit to be able to [create mutable refs when it syncs "local" sources](https://github.com/moby/buildkit/blob/master/source/local/local.go#L127), but also to enable the ability of generating layers with arbitrary files that could be exported and imported again. These layers are not necessarily usable as layers from which we can "boot" a container.

This work has a hcsshim component that creates a number of files needed for a layer to be considered valid. The layout of the layer requires that there is a ```Files``` folder, which becomes the ```rootfs``` of the container, if this layer were to be "bootable". Currently, ```hcsshim``` and ```HCS``` in general doesn't have any other kind of layer it supports, but ```WCIFS```, so we create a valid ```WCIFS``` layer to "trick" it.

## The containerd snapshotter

The code proposed here, leverages that new functionality in ```hcsshim``` and enables the creation of new, parent-less base layers. The code currently assigns a new mount type called ```bind``` for these parent-less layers, and mutates the source to the ```Files``` folder.

The client can then choose to mount this new ```mount.Mount{}``` at the desired location using ```m.Mount(target string)```.

## Mounting a ```mount.Mount{}```

### Windows layers

On Windows, before this PR, the only mount type supported was ```windows-layer```. Before a windows layer can be mounted on the host, we need to "activate" it, which creates a new ```\\?\Volume{}``` via ````WCIFS```. At least for writable layers. This new volume is a "gateway" to the ```Files``` folder. We can mount it on a host and mutate it if we need to.

For parent-less base layers, ```ActivateLayer()``` and ```PrepareLayer()``` don't really create a new WCIFS volume. Calling ```GetLayerMountPath()``` on an activated parent-less base layer, will return the actual snapshot location.

 ### The issue of "bind" mounts

Bind mounts, as they are introduced here, are a special case, and in the case of Windows, may be useful, but only in the context of a "local" mounter. The shim cannot handle "bind" mounts. So we have a choice to either "support" this in containerd when mounting locally, or just do away with it. We could in theory simply [leave the if statement](https://github.com/containerd/containerd/pull/8043/files#diff-ede639814a8abed289b5808723cc84b4aeee754dfc97fbaa435fb0dc4e910b7dR52-R57), and it should be fine, but that is about it in terms of "bind" support.

## Suggested course of action

Given the above, here is what I suggest:

  * The snapshotter should not concern itself with any kind of layers besides ```windows-layer```.
    * After creating a new parent-less layer, the [mounts()](https://github.com/containerd/containerd/pull/8043/files#diff-b6ae335f862c747466f1bc00e89eb1ef7cd9589bcc89ec57953c9c20125b3752R306) function should **not** concern itself with the mount type. It will remain a ```windows-layer```, but we will however check for the existence of parent layers and only add ```parentLayersOption``` if there are any parent layers.
    * We will no longer append ```Files``` to source, but simply return it as is.
  * In [mount.mount()](https://github.com/containerd/containerd/pull/8043/files#diff-ede639814a8abed289b5808723cc84b4aeee754dfc97fbaa435fb0dc4e910b7dR102), right before ```ApplyFileBinding()```, we check if this layer has any parents. In the case of a parent-less layer, [the volume](https://github.com/containerd/containerd/pull/8043/files#diff-ede639814a8abed289b5808723cc84b4aeee754dfc97fbaa435fb0dc4e910b7dR97) will not be a WCIFS ```\\?\Volume{}```, but the actual path on disk to the snapshot location. We cannot mount this, as a client like buildkit will potentially remove everything before writing something to the target. This, for example, actually happens due to how ```diffcopy``` works in fsutil. Source and target are synced something like how ```rsync -a --delete``` works.
    * In case of a WCIFS volume, we simply mount it, as it already is a gateway inside ```Files```
    * In the case of a parent-less base layer, we check for the existence of ```Files``` and mount that inside the target.

The local mounter in buildkit will be amended. Right now it mounts layers in-place, something that is not necessary. We also allow ```Mount()``` to do the right thing, instead of returning the source if it's not ```ro```.

What do you think?

gabriel-samfira replied:
I've pushed: https://github.com/containerd/containerd/pull/8043/commits/d373ebc4de9a514af7d42c2edc980ec1d0868598

gabriel-samfira replied:
Marking this as resolved. Let me know if you have any thoughts on the above or if you can think of anything I've missed.

fuweid commented on mount/mount_windows.go:
```diff
@@ -17,17 +17,45 @@
+	ErrNotImplementOnWindows = errors.New("not implemented under windows")
+)
+
+func (m *Mount) ReadOnly() bool {
```
I think we should not export this one because we don't export this in other platform.

gabriel-samfira replied:
Will change this tomorrow morning.

gabriel-samfira replied:
Would it be acceptable if I export it for all platforms? It simply looks for the ```ro``` option in a string array. If it's too specific, I could have it accept a parameter and rename it to ```HasOption()``` or similar. What do you think?

gabriel-samfira replied:
Moved it to ```mount.go```. It's currently being used in the ```mount``` package and in the ```diff``` package. Let me know if you prefer another solution.

fuweid reviewed (commented):
LGTM

I would like to wait for more eyes on this one. Thanks

gabriel-samfira commented on diff/windows/windows.go:
```diff
@@ -332,6 +335,22 @@ func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
 		return "", nil, err
 	}
 
+	if mnt.ReadOnly() {
```
@TBBle given the changes in https://github.com/containerd/containerd/commit/d373ebc4de9a514af7d42c2edc980ec1d0868598, do you think the changes in this file are needed anymore?

TBBle replied:
We can definitely get rid of _some_ of the changes (the ones that were caring about "bind", since we don't generate that type anymore), but the logic here might need revision as I _believe_ it was based on the old code where a View of a base layer still generated a scratch layer. (So I suspect it's not working currently here...)

It's possible that the old code worked fine. The case to watch out for is the one in that comment about `rootfs.CreateDiff`, as that code-path can pass a lower mount created by `sn.View(ctx, lowerKey, "")`, i.e. a read-only mount of a layer containing just a "Files" directory, and we need to ensure we trigger the base-layer exporter in hcsshim, hence the code in `mountsToLayerAndParents` that recognises such a mount and returns magic `""`, and the code in `mountPairToLayerStack` that recognises a lower of `""` and an upper with no parents, and returns just the upper.

This code-path is new with this PR, since until this PR, the Windows Snapshotter couldn't diff/export a base layer, it could only import them.

I'm not sure if any test code exercises that code-path, as it's mostly used by the checkpointing code that I suspect doesn't work in Windows for other reasons. But `ctr diff` also calls `rootfs.CreateDiff`, so I expect you could test the problematic code path by trying to `ctr diff` a base layer.

So my guess would be we still need _some_ changes in this file, since the existing code checked for the "new empty view" case assuming the snapshotter returned an empty list of mounts, apparently. But they can be simpler because we no longer have the case where we have to check for "bind" mount with parents, or a "windows-layer" mount without parents.

We definitely need to strip off the dummy sandbox on a non-base-layer View, otherwise the diff would be empty as we'd pass the dummy sandbox into hcsshim.

So  `mountsToLayerAndParents` only needs to keep the new `mnt.ReadOnly` code-path, and that can be simplified to something like:
```
	if mnt.ReadOnly() {
		if len(parentLayerPaths) == 0 {
			// A view without parents is a placeholder for "empty". We special-case this to simplify the caller.
			return "", nil, nil
		}
		// A view with parents points to a trivial scratch layer parented to the layer of interest, so strip it off.
		return parentLayerPaths[0], parentLayerPaths[1:], nil
	}
```

Then the logic in `mountPairToLayerStack` should still work; the main change this PR made was to look for that `""` as the lower layer path instead of an empty Mount array. I don't think it was ever possible to get an empty array there though (the Windows snapshotter always returns an error or a single Mount), so I suspect that code had never worked.

And a note for testing, `ctr diff` cannot diff an imported non-base layer, as HCS's diff export uses the sandbox.vhdx of the upper layer, and that only exists if a layer was created locally and committed. (Before the optimisation to skip sandbox.vhdx creation on imports, this case would produce an empty diff; now it should fail, again down in hcsshim).

So you'll have to create and populate the layers for testing locally.

gabriel-samfira replied:
So, here's what I tried:

![Screenshot from 2023-04-03 17-46-30](https://user-images.githubusercontent.com/903160/229545194-aa1a9629-d2d8-4fd9-a517-36c8445bea02.png)

The ```testing``` snapshot was created by buildkit before trying to upload  ```.dockerignore``` which didn't exist, so it is empty. I then created a new active snapshot from it. Created a diff of the ```testing``` snapshot (has no parents), and one of the ```2``` snapshot.

Seems to do the right thing. This is with the changes you mentioned above.

gabriel-samfira replied:
creating a diff of a layer that resulted from an image import seems to work as before. So does creating a diff of a child snapshot from that image.

If you see anything that I've missed, let me know.

gabriel-samfira commented:
Hi @dmcgowan @estesp @kevpar 

Could you spare a few minutes to look over this PR?

kzys commented on snapshots/testsuite/issues.go:
```diff
@@ -94,6 +95,9 @@ func checkRemoveDirectoryInLowerLayer(ctx context.Context, t *testing.T, sn snap
 // see https://github.com/docker/docker/issues/28391 overlay2
 func checkChown(ctx context.Context, t *testing.T, sn snapshots.Snapshotter, work string) {
+	if runtime.GOOS == "windows" {
+		t.Skip("Chown is not supported on WCOW")
```
Nit: maybe just "... not supported on Windows"? Even LCOW, it is not supported on the Windows side.

kzys commented on snapshots/testsuite/testsuite.go:
```diff
@@ -869,6 +876,10 @@ func closeTwice(ctx context.Context, t *testing.T, snapshotter snapshots.Snapsho
 
 func checkRootPermission(ctx context.Context, t *testing.T, snapshotter snapshots.Snapshotter, work string) {
+	if runtime.GOOS == "windows" {
+		t.Skip("Filesystem permissions are not supported on WCOW")
```
Like above, WCOW -> Windows

kzys commented on snapshots/windows/windows.go:
```diff
@@ -349,73 +352,80 @@ func (s *snapshotter) createSnapshot(ctx context.Context, kind snapshots.Kind, k
 		snDir := s.getSnapshotDir(newSnapshot.ID)
 		if err = os.MkdirAll(snDir, 0700); err != nil {
-			return err
+			return fmt.Errorf("creating snapshot dir: %w", err)
```
`failed to create`?

Also if `err` doesn't mention the path, it would be better to include snDir in this error.

gabriel-samfira replied:
This was a consequence of bouncing between ```errors.Wrap()``` and ```fmt.Errorf()``` :smile:. Will fix.

TBBle replied:
I'm pretty sure `os.MkdirAll` always returns a `PathError`. Cheating and checking the source says it either returns `PathError` or calls `os.Mkdir`, and _it_ is documented to always return a `PathError`.

gabriel-samfira replied:
added the snDir for good measure.

kzys reviewed (approved):
Looks good to me. Only have nits and I'm fine addressing them later.

gabriel-samfira commented:
Thank you @kzys !

Addressing the nits now.

TBBle commented on mount/mount_windows.go:
```diff
@@ -132,14 +132,6 @@ func (m *Mount) GetParentPaths() ([]string, error) {
-
-		return fmt.Errorf("failed to access mount point %s: %w", mount, err)
-	}
-
```
Why was this block removed? AFAIR it's necessary because `Unmount("/nosuchpath", 0)` is valid and supposed to be a no-op. Or are we just relying on `bindfilter.RemoveFileBinding` to implement that?

gabriel-samfira replied:
I added it previously in https://github.com/containerd/containerd/pull/8043/commits/6a5b4c9c242a0278033b20f7f5eca4fdbbea231a. It felt wrong to mask an error that the target is not there. On Linux, this function ends up calling ```unix.Unmount``` which will err out if the target does not exist.

Should I leave it in?

TBBle replied:
I think we crossed edits. Linux `Unmount` doesn't ignore this case, but `UnmountAll` does. My initial comment was confused with `""` handling which we already have in Windows `UnmountAll`.

gabriel-samfira replied:
in that case I think we can move that check in ```UnmountAll()```. The bind filter will probably not return an error that matches ```IsNotExist```.

gabriel-samfira replied:
Ok. Same thing on linux: https://github.com/containerd/containerd/blob/main/mount/mount_linux.go#L181-L183

gabriel-samfira replied:
added https://github.com/containerd/containerd/pull/8043/commits/c9e5c33a18acffe583592828a5a6696cbc779c1a

MikeZappa87 commented:
/lgtm

dcantah commented:
Anyone able to check this in? We have 4 approvals from maintainers

TBBle commented:
Woo, and I cannot emphasise this enough, hoo!

gabriel-samfira commented:
> Woo, and I cannot emphasise this enough, hoo!

There are [insert_favorite_beverage] I need to send your way :smile:.

fuweid commented:
@gabriel-samfira @TBBle thanks for making this and your patience.

dcantah commented:
Thrilled to see this land finally 🥳
